os/kernel/semaphore/sem_holder.c

```diff
@@ -908,11 +908,8 @@ void sem_releaseholder(FAR sem_t *sem, FAR struct tcb_s *rtcb)
 		 */
 		pholder->counts--;
 
-#ifndef CONFIG_PRIORITY_INHERITANCE
-		if (pholder->counts <= 0) {
-			sem_freeholder(sem, pholder);
-		}
-#endif
+		/* Remove the holder from the holder list */
+		sem_freeholder(sem, pholder);
```

Comment:
I think we need
`#if !defined(CONFIG_PRIORITY_INHERITANCE) && defined(CONFIG_BINMGR_RECOVERY)`
and we should check the pholder counts.
Because pholder counts can be more than 0 when same task tries to wait the sem in normal case.